### PR TITLE
[JENKINS-30521] reload-job CLI command moved from Core to CLI

### PR DESCRIPTION
**Pull request by pjanouse**
Please review, thx.

### CODE REVIEW

File: core/src/main/java/hudson/cli/ReloadJobCommand.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2013-5 Red Hat, Inc.
```
Comment (KostyaSha) [inline core/src/main/java/hudson/cli/ReloadJobCommand.java]: 2013-5 for new file?

File: test/src/test/java/hudson/cli/DeleteJobCommandTest.java [hunk 2]
```diff
@@ -33,7 +33,9 @@
 
 import java.io.IOException;
 
-import static hudson.cli.CLICommandInvoker.Matcher.*;
+import static hudson.cli.CLICommandInvoker.Matcher.failedWith;
```
Comment (KostyaSha) [inline test/src/test/java/hudson/cli/DeleteJobCommandTest.java]: unrelated

File: core/src/main/resources/hudson/cli/Messages_pt_BR.properties [hunk 3]
```diff
@@ -114,4 +114,7 @@ InstallPluginCommand.ShortDescription=Instala um plugin a partir de um arquivo,
 # Deletes a node
 CLI.delete-node.shortDescription=Remover o n\u00f3
 # Deletes a job
-DeleteJobCommand.ShortDescription=Remover uma job
+DeleteJobCommand.ShortDescription=Remover uma job(s)
+
+ReloadJobCommand.ShortDescription=\
```
Comment (KostyaSha) [inline core/src/main/resources/hudson/cli/Messages_pt_BR.properties]: why it need to be on new line?

**Comment by pjanouse** [comment]
Fixed

**Comment by oleg-nenashev** [comment]
Is this migration transparent for existing user scripts?

File: test/src/test/java/hudson/cli/ReloadJobCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+                .authorizedTo(Job.READ, Job.CONFIGURE, Jenkins.READ)
+                .invokeWithArgs("aProject1", "aProject2", "aProject1");
+
+        assertThat(result, succeededSilently());
+
+        assertThat(project1.scheduleBuild2(0).get().getLog(), containsString("echo 2"));
+        assertThat(project2.scheduleBuild2(0).get().getLog(), containsString("echo 2"));
+    }
+
+    private void changeProjectOnTheDisc(final FreeStyleProject p, final String oldstr,
```
Comment (olivergondza) [inline test/src/test/java/hudson/cli/ReloadJobCommandTest.java]: Nit: the implementation can be simplified using `FilePath#readToString`/`FilePath#write(String, String)`.

**Comment by olivergondza** [comment]
@pjanouse , Oleg is right. there seems to be a change in functionality. See:

```
git checkout eb4669540a19a095aca2475435ebb22032ef134c~1
git checkout eb4669540a19a095aca2475435ebb22032ef134c */src/test/
mvn clean package -Dtest=ReloadJobCommandTest -DfailIfNoTests=false
```
(please use `CLICommandInvoker(JenkinsRule, String)` for such refactorizations so this is easy to verify).

There are some interesting failures aside from "Too many arguments" that are (caused by the fact you started to support bulk reload and thus) harmless.

**Comment by olivergondza** [comment]
Also this seems to be a bug in the original implementation:

```
java.lang.NullPointerException
	at hudson.model.AbstractItem.resolveForCLI(AbstractItem.java:727)
              ...
```

**Comment by pjanouse** [comment]
I can confirm the occurred NPE is a bug, tracked in JENKINS-30629 + PR sent already.

I accept other notes and working on them currently. Thank you.

**Comment by pjanouse** [comment]
Fixed, PR updated.

Note: Current implementation depends on https://github.com/jenkinsci/jenkins/pull/1841

**Comment by pjanouse** [comment]
Updated due to https://github.com/jenkinsci/jenkins/pull/1841

**Comment by pjanouse** [comment]
Rebased...

**Comment by pjanouse** [comment]
Fixed error message for a backward compatibility (when project doesn't exist).

File: test/src/test/java/hudson/cli/ReloadJobCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+import static org.hamcrest.Matchers.containsString;
+
+/**
+ * @author pjanouse
+ */
+public class ReloadJobCommandTest {
+
+    private CLICommandInvoker command;
+
+    @Rule public final JenkinsRule j = new JenkinsRule();
```
Comment (oleg-nenashev) [inline test/src/test/java/hudson/cli/ReloadJobCommandTest.java]: Not important: Just for the future, it's not the standard code style in Jenkins. The core uses [Oracle code style](http://www.oracle.com/technetwork/java/javase/documentation/codeconvtoc-136057.html). BTW @jglick uses this one.

File: core/src/main/java/hudson/cli/ReloadJobCommand.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        final HashSet<String> hs = new HashSet<String>();
+        hs.addAll(jobs);
+
+        for (String job_s: hs) {
+            AbstractItem job = null;
+
+            try {
+                job = (AbstractItem) jenkins.getItemByFullName(job_s);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/ReloadJobCommand.java]: Unchecked Cast. I suppose you should use <code>Item</code> here (there are no other public implementations, but they may exist in closed source plugins).

File: core/src/main/java/hudson/cli/ReloadJobCommand.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+                if(job == null) {
+                    AbstractProject project = AbstractProject.findNearest(job_s);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/ReloadJobCommand.java]: Why don't you check other job types? The current implementation does not support Workflow jobs, Inheritance templates, etc.

**Comment by oleg-nenashev** [comment]
The code mostly looks good to me. BTW, there is so erroneous Item type handling logic => keeping :-1: for the current state

File: core/src/main/java/hudson/cli/ReloadJobCommand.java (hunk 6 above)
Comment (pjanouse) [inline core/src/main/java/hudson/cli/ReloadJobCommand.java]: But Item (Interface) doesn't declare any useful 'reload' behavior. I don't want to duplicate a code from AbstractProject.doReload() here. 

Nobody can rely on reload from Item because it doesn't exists - so I can't break a backward compatibility as well.

File: core/src/main/java/hudson/cli/ReloadJobCommand.java (hunk 7 above)
Comment (pjanouse) [inline core/src/main/java/hudson/cli/ReloadJobCommand.java]: Because I've used that for backward compatibility only. 

For me it is enough to say 'Job X doesn't exist', but it can break a backward compatibility if somebody rely on the hint, so I've used the same code as in https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/AbstractItem.java#L727-L729 (I did a fix there last week)

Original reload-job CLI command was defined over AbstractItem, so I think we should do here the same.

File: core/src/main/java/hudson/cli/ReloadJobCommand.java (hunk 6 above)
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/ReloadJobCommand.java]: Then use something like that:
```java
Item item = jenkins.getItemByFullName(job_s);
if (item instanceof AbstractItem) {
    job = (AbstractItem) item;
} elseif (item != null) {
    LOGGER.log(WARNING, "Unsupported item type: {0}", item.getClass().getName())
}
```

File: core/src/main/java/hudson/cli/ReloadJobCommand.java (hunk 7 above)
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/ReloadJobCommand.java]: @pjanouse 
<code>AbstractProject</code> != <code>AbstractItem</code>. BTW I agree that this issue has been caused by the original code. We could create a follow-up bug, add TODO to the code and do not change the behavior now.

**Comment by pjanouse** [comment]
+1
Fixed and merged

**Comment by oleg-nenashev** [comment]
:+1:
